### PR TITLE
Tutorial notebook for Lagrangian Neural Networks

### LINKED ISSUE #2680: DeepChem Model Wishlist
There are a number of important molecular and scientific models that I would love to see implementations for in DeepChem. Implementations of any of these models would be a great contribution to DeepChem. As an incentive for contributors, contributing a model would earn authorship on one of our upcoming manuscripts! If you have bandwidth to contribute an implementation, please chime in below 

- [ ] IterRefLSTMs from https://pubs.acs.org/doi/10.1021/acscentsci.6b00367 (We have some old DeepChem code that doesn't work in today's DeepChem that can be ported, see https://github.com/deepchem/deepchem/issues/2070). 
- [x] D-MPNN (from Chemprop, https://github.com/chemprop/chemprop, https://pubs.acs.org/doi/abs/10.1021/acs.jcim.9b00237). @atreyamaj has volunteered to take a crack at the implementation
- [x] Grover (from https://github.com/tencent-ailab/grover, https://papers.nips.cc/paper/2020/hash/94aef38441efa3380a3bed3faf1f9d5d-Abstract.html). Current state of the art on MoleculeNet
- [x] FermiNet (from https://github.com/deepmind/ferminet, https://deepmind.com/blog/article/FermiNet)
- [ ] PauliNet (from https://github.com/deepqmc/deepqmc, https://arxiv.org/abs/1909.08423)
- [ ] MXMNet (from https://github.com/zetayue/MXMNet, https://arxiv.org/abs/2011.07457). Current state of the art on QM9
- [x] MEGNet (from https://pubs.acs.org/doi/abs/10.1021/acs.chemmater.9b01294, https://github.com/materialsvirtuallab/megnet)
- [ ] MultiScaleGNN (from https://arxiv.org/pdf/2106.04900.pdf)
- [ ] Rexgen (from https://github.com/connorcoley/rexgen_direct, https://pubs.rsc.org/en/content/articlelanding/2019/sc/c8sc04228d)
- [ ] Retrotemp (from https://github.com/connorcoley/retrotemp, https://chemistry-europe.onlinelibrary.wiley.com/doi/abs/10.1002/chem.201605499)
- [ ] DeepONet (from https://github.com/lululxvi/deeponet, https://github.com/lululxvi/deepxde)
- [ ] Universial DIfferential Equations (from https://arxiv.org/pdf/2001.04385.pdf, https://github.com/ChrisRackauckas/universal_differential_equations)
- [ ] Neural ODEs (from https://github.com/rtqichen/torchdiffeq, https://github.com/msurtsukov/neural-ode, https://arxiv.org/abs/1806.07366)
- [ ] Lagrangian Neural Networks (from https://arxiv.org/abs/2003.04630, https://github.com/MilesCranmer/lagrangian_nns)
- [ ] Hamiltonian Neural Networks (from https://arxiv.org/abs/1906.01563, https://github.com/greydanus/hamiltonian-nn)
- [ ] hPINNs (from https://arxiv.org/abs/2102.04626, https://github.com/lululxvi/deepxde)
- [x] Gaussian Processes (this is already in scikit learn https://scikit-learn.org/stable/modules/gaussian_process.html, so it would be sufficient to add a tutorial. Might be good for beginners)
- [ ] NequIP (from https://arxiv.org/abs/2101.03164, https://github.com/mir-group/nequip)
- [ ] Junction Tree Variational Autoencoders (https://github.com/wengong-jin/icml18-jtnn, https://arxiv.org/abs/1802.04364)
- [ ] Convolutional Neural ODEs (https://ml4physicalsciences.github.io/2020/files/NeurIPS_ML4PS_2020_125.pdf)

## Equivariant Models
- [ ] Tensor Field Network (https://github.com/tensorfieldnetworks/tensorfieldnetworks)
- [ ] E3NN (from https://github.com/e3nn/e3nn, https://arxiv.org/abs/2101.03164)
- [ ] SE(3) Transformer (from https://github.com/FabianFuchsML/se3-transformer-public, https://github.com/lucidrains/se3-transformer-pytorch, https://arxiv.org/abs/2006.10503)

## Classical Machine Learning
- [ ] Hidden Markov Models (perhaps just a wrapper around https://github.com/hmmlearn/hmmlearn). 
- [ ] Kalman Filter (can use https://github.com/rlabbe/filterpy as a reference.)

### PR DESCRIPTION
**Pull request by PsVenom**
## Description

Fix #2680 

Added a self-contained tutorial for Lagrangian Neural Networks 

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by PsVenom** [comment]
<p>Hey shreyas, this rendering issue seems to emerge from reviewnb's UI. </p>

---

 <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/#comment-2007605102'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by PsVenom** [comment]
<p>done, please check</p>

---

 <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/#comment-2007605072'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by PsVenom** [comment]
I've removed the JAX tutorial and added a Pytorch port

File: examples/tutorials/Introduction to Langrangian Neural Networks.ipynb [hunk 1]
```diff
@@ -0,0 +1,824 @@
+{
```
Comment (shreyasvinaya) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Add titles and some information on what the graphs mean</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Equation not rendering properly with text</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Spell check and make this part slightly more descriptive. Maybe show how you got the equations for r and theta if it's not too lengthy</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Explain what these functions do, what are the inputs, what is the output</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Rename the functions, make the names intuitive and simple but descriptive. Add docstrings to explain what the functions do, 1-2 lines will do</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #27.        def f_anal(x, t):
</pre><p>From the indentation, it looks like you're defining a function within a function</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    import time
</pre><p>Avoid importing packages mid cell, do it the top to make things simple</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    print(end)
</pre><p>What is the output? Maybe add some text alongside the output explaining what it is</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Explain what this function does</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Put this cell at the top of the notebook, near the importing of packages. You've used <code>.detach()</code> in above cells, it's better to keep the device constant at the start, instead of switching mid tutorial </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #46.    #setting the lagrangian loss function (in the original paper, it was set to be the L2 distance between the original and predicted outcomes)
</pre><p>I belive Deepchem has L2 loss as a function, try to import and use that here instead of defining the loss function again</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (aaronrockmenezes) [inline examples/tutorials/Introduction to Langrangian Neural Networks.ipynb:1]: <p>Train model for a lesser number of epochs rather than stopping training mid way. Loss seems to reduce well</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deepchem/deepchem/pull/3905/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>
